src/rules/preferReadonlyRule.ts

```diff
@@ -0,0 +1,189 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+import { ClassScope, ParameterOrPropertyDeclaration } from "./prefer-readonly/classScope";
+
+export class Rule extends Lint.Rules.TypedRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Requires that private variables are marked as `readonly` if they're never modified outside of the constructor.",
+        descriptionDetails: Lint.Utils.dedent`
+            If a private variable is only assigned to in the constructor, it should be declared as \`readonly\`.
+        `,
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        requiresTypeInfo: true,
+        ruleName: "prefer-readonly",
+        type: "maintainability",
+        typescriptOnly: true,
+    };
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        const preferReadonlyWalker = new PreferReadonlyWalker(sourceFile, this.ruleName, program.getTypeChecker());
+        return this.applyWithWalker(preferReadonlyWalker);
+    }
+}
+
+class PreferReadonlyWalker extends Lint.AbstractWalker<void> {
+    private readonly typeChecker: ts.TypeChecker;
+    private scope?: ClassScope;
+
+    public constructor(sourceFile: ts.SourceFile, ruleName: string, typeChecker: ts.TypeChecker) {
+        super(sourceFile, ruleName, undefined);
+
+        this.typeChecker = typeChecker;
+    }
+
+    public walk(sourceFile: ts.SourceFile) {
+        if (sourceFile.isDeclarationFile) {
+            return;
+        }
+
+        ts.forEachChild(sourceFile, this.visitNode);
+    }
+
+    private readonly visitNode = (node: ts.Node) => {
+        switch (node.kind) {
+            case ts.SyntaxKind.ArrowFunction:
+            case ts.SyntaxKind.FunctionExpression:
```

Comment:
There are a few more kinds to check here: `ClassDeclaration`, `ClassExpression`, `FunctionDeclaration`, `GetAccessor`, `SetAccessor`, `MethodDeclaration` and maybe some more:
http://www.typescriptlang.org/play/#src=class%20Foo%20%7B%0A%20%20%20%20private%20readonly%20foo%3A%20number%3B%0A%0A%20%20%20%20constructor()%20%7B%0A%20%20%20%20%20%20%20%20this.foo%20%3D%202%3B%0A%20%20%20%20%20%20%20%20let%20self%20%3D%20this%3B%0A%20%20%20%20%20%20%20%20class%20Inner%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20prop%20%3D%20self.foo%2B%2B%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20let%20innerObj%20%3D%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20doStuff()%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2B%2Bself.foo%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%2C%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20get%20foo()%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20return%20self.foo%2B%2B%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%2C%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20set%20foo(foo)%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20self.foo%20%3D%20foo%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%0A%20%20%20%20%20%20%20%20function%20fn()%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20self.foo%2B%2B%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%7D